lib/project_types/script/layers/infrastructure/script_project_repository.rb

```diff
@@ -140,40 +137,110 @@ def validate_metadata!(extension_point_type, language)
+          @script_config_repository ||= ScriptConfigRepository.new(ctx: ctx)
+        end
+
+        class ScriptConfigRepository
           include SmartProperties
           property! :ctx, accepts: ShopifyCLI::Context
 
-          def get
-            current_script_json || raise(Domain::Errors::NoScriptJsonFile)
+          def get!
+            yml_repository.get || json_repository.get || raise(Infrastructure::Errors::NoScriptConfigYmlFileError)
           end
 
           def update_or_create(title:)
-            json = current_script_json&.content || {}
-            json["version"] ||= "1"
-            json["title"] = title
+            json_repository.update(title: title) || yml_repository.update_or_create(title: title)
+          end
 
-            ctx.write(SCRIPT_JSON_FILENAME, JSON.pretty_generate(json))
+          private
 
-            Domain::ScriptJson.new(content: json)
+          def yml_repository
+            @yml_repository ||= ScriptConfigYmlRepository.new(ctx: ctx)
           end
 
-          private
+          def json_repository
+            @json_repository ||= ScriptJsonRepository.new(ctx: ctx)
+          end
 
-          def current_script_json
-            return nil unless ctx.file_exist?(SCRIPT_JSON_FILENAME)
+          class ScriptConfigYmlRepository
+            include SmartProperties
+            property! :ctx, accepts: ShopifyCLI::Context
+
+            SCRIPT_CONFIG_YML_FILENAME = "script.config.yml"
+
+            def get
+              return nil unless ctx.file_exist?(SCRIPT_CONFIG_YML_FILENAME)
+
+              content = ctx.read(SCRIPT_CONFIG_YML_FILENAME)
+              require "yaml"
+              begin
+                hash = YAML.load(content)
+              rescue Psych::SyntaxError
+                raise Errors::InvalidScriptConfigYmlDefinitionError
+              else
+                raise Errors::InvalidScriptConfigYmlDefinitionError unless hash.is_a?(Hash)
+                from_h(hash)
+              end
+            end
+
+            def update_or_create(title:)
+              hash = get&.content || {}
+              hash["version"] ||= "1"
+              hash["title"] = title
+
+              ctx.write(SCRIPT_CONFIG_YML_FILENAME, YAML.dump(hash))
 
-            content = ctx.read(SCRIPT_JSON_FILENAME)
-            raise Domain::Errors::InvalidScriptJsonDefinitionError unless valid_script_json?(content)
+              from_h(hash)
+            end
+
+            private
 
-            Domain::ScriptJson.new(content: JSON.parse(content))
+            def from_h(hash)
+              Domain::ScriptConfig.new(content: hash)
+            rescue Domain::Errors::MissingScriptConfigFieldError => e
+              raise Errors::MissingScriptConfigYmlFieldError, e.field
+            end
           end
 
-          def valid_script_json?(content)
-            JSON.parse(content)
-            true
-          rescue JSON::ParserError
-            false
+          class ScriptJsonRepository
```

Comment:
Just took a very quick look, and I think some of the code is just temporary, so, please don't feel obligated to change anything.  I do think the relationship is not that ScriptConfigRepo has ScriptJsonRepo and ScriptConfigYmlRepo.  I think it's really ScriptJsonRepo/ScriptConfigYmlRepo is a ScriptConfigRepo.  So, I think the inheritance relationship would be better here, with consistent contract in ScriptConfigRepo as the parent.

Reply:
The reason that I didn't use inheritance was because the logic in the `ScriptConfigRepo` uses both the JSON and YML repo [in its instance methods](https://github.com/Shopify/shopify-cli/blob/7a44db91bd17cebd03ecd9698060bf9efe93b29a/lib/project_types/script/layers/infrastructure/script_project_repository.rb#L148-L154), and I don't think it is good practice for a base class to know about the subclasses in instance methods. If we did want to use inheritance, I suppose we could move the logic that is currently in `ScriptConfigRepository`'s instance methods to a separate class or perhaps instance methods on `ScriptProjectRepository`. WDYT?

Reply:
If we wanted to keep this logic long term, then I would agree that inheritance makes more sense here. Since we have a stretch goal to remove support for `script.json` next cycle, I personally think what we have here would be fine for now. But I can give my thoughts if we want to bikeshed in the meantime 😄 

> I don't think it is good practice for a base class to know about the subclasses in instance methods

+1 I definitely agree! I think both your alternative ideas could work. Since the `ScriptProject` is the aggregate root over the script config files, I think it would make sense to leave logic in the `ScriptProjectRepo` as long as it isn't too noisy (which i don't think it would be since its rather minimal).

Reply:
Thanks for the detailed answer! I applied a lot of your suggestion to the `get` method, which is now only implemented in the base class. I think the updating is more nuanced with the `ScriptConfigYmlRepository` having an `update_or_create` method and the `ScriptJsonRepository` only having an `update` method so I did not use inheritance on those methods.